[IR][JumpThreading] Fix infinite recursion on compare self-reference [updated]

ro-i opened the pull request:
EDIT March 21: updated title. See commits and current discussion for the current state of this PR.

In case of an unreachable loop, for example, the JumpThreading pass might attempt to remove the PHI node in the loop entry and replace it with its constant value. However, this fails in case the value is an instruction using the PHI node.

This LLVM defect was identified via the AMD Fuzzing project.

PS: I felt a bit discouraged from fixing this on the basic block level, but I think that the added check only verifies what should hold anyway.

ro-i committed:
[IR] Avoid self-referencing values caused by PHI node removal

In case of an unreachable loop, for example, the JumpThreading pass
might attempt to remove the PHI node in the loop entry and replace it
with its constant value. However, this fails in case the value is an
instruction using the PHI node.

This LLVM defect was identified via the AMD Fuzzing project.

arsenm commented on llvm/test/Transforms/JumpThreading/unreachable-loops.ll:
```diff
@@ -180,4 +180,64 @@ cleanup2343.loopexit4:                            ; preds = %cleanup1491
+  br i1 %L10, label %BB6, label %F6
+
+BB6:                                              ; preds = %BB5
+  %LGV3 = load i1, ptr null, align 1
```
Avoid the undefined null load

ro-i replied:
llvm-reduce changed that ^^ ig because it thought that a proper stack slot, which was there before, isn't necessary for the test case? But let me change it back

arsenm replied:
can use an argument or global or anything, it doesnt' really matter

arsenm commented on llvm/test/Transforms/JumpThreading/unreachable-loops.ll:
```diff
@@ -180,4 +180,64 @@ cleanup2343.loopexit4:                            ; preds = %cleanup1491
+BB6:                                              ; preds = %BB5
+  %LGV3 = load i1, ptr null, align 1
+  %C4 = icmp sle i1 %L6, true
+  store i1 %C4, ptr null, align 1
```
Avoid undefined null store?

ro-i replied:
(see my comment above)

arsenm replied:
llvm-reduce will introduce them but you should try to remove them after

ro-i committed:
implement feedback

nikic reviewed (changes requested):
It is legal to have self-referential instructions in unreachable code. If JumpThreading can't handle the result, that needs to be fixed, not the creation of the self-reference.

ro-i committed:
implement another approach

ro-i commented:
I added another implementation. This checks much more aggressively for blocks that become unreachable during pass execution. Consequently, I had to update a few of the existing tests because their checks contained unreachable blocks, which get deleted now by the pass. (If you agree that this is a viable solution, I may further need to adapt some of these tests because right now, they may not check for their actual test case anymore.

arsenm commented on llvm/lib/Transforms/Scalar/JumpThreading.cpp:
```diff
@@ -335,15 +335,19 @@ bool JumpThreadingPass::runImpl(Function &F_, FunctionAnalysisManager *FAM_,
-        DeleteDeadBlock(&BB, DTU.get());
+        for (BasicBlock *UBB : Unreachable) {
+          LLVM_DEBUG(dbgs()
+                     << "  JT: Deleting dead block '" << UBB->getName()
```
Should use getNameOrAsOperand

ro-i committed:
replace undef

arsenm commented on llvm/lib/Transforms/Scalar/JumpThreading.cpp:
```diff
@@ -381,6 +385,26 @@ bool JumpThreadingPass::runImpl(Function &F_, FunctionAnalysisManager *FAM_,
 }
 
+SmallVector<BasicBlock *, 32>
+JumpThreadingPass::unreachableFromBB(BasicBlock *BB, DomTreeUpdater *DTU) {
```
This feels like a heavy hammer, than you also shouldn't need to reimplement. Can you just skip any block where !DT.isReachableFromEntry?

ro-i replied:
Let me see if it's possible to make the pass track new unreachable blocks correctly using its `Unreachable` container

ro-i committed:
implement another approach; reset existing tests

ro-i commented:
Ok, I now created a third approach :) Let me first show the situation more in detail:
```
define i32 @constant_phi_leads_to_self_reference(ptr %ptr) {
  %A9 = alloca i1, align 1
  br label %F6
 
BB4:                                              ; No predecessors!
  unreachable
 
F1:                                               ; No predecessors!
  unreachable
 
T3:                                               ; preds = %T4, %BB6
  %L6 = phi i1 [ %L6.pr, %T4 ], [ %C4, %BB6 ]
  br label %BB5
 
BB5:                                              ; preds = %F7, %T3
  %L10 = load i1, ptr %A9, align 1
  br i1 %L10, label %BB6, label %F6
 
BB6:                                              ; preds = %BB5
  %LGV3 = load i1, ptr %ptr, align 1
  %C4 = icmp sle i1 %L6, true
  store i1 %C4, ptr %ptr, align 1
  br i1 %L6, label %F6, label %T3
 
T4:                                               ; No predecessors!
  %L6.pr = load i1, ptr %ptr, align 1
  br label %T3
 
F6:                                               ; preds = %0, %BB6, %BB5
  ret i32 0
 
F7:                                               ; No predecessors!
  br label %BB5
}
```
This is the state of the function before the problem occurs while handling T4. Why is T4 a problem? It's the pre-header of T3 (which is a loop header). L6.pr is used by L6. The pass now detects that T4 is unreachable and removes it. Consequently, L6 become `phi i1 [%C4, %BB6]`. At a later step, this constant phi node will then be simplified. As a result, every use of L6 will be replaced by C4, which leads to a self-reference in C4.
IMHO, there are 3 good approaches to handle this problem:
1. Avoid creating invalid IR at a general level. This was the first approach of my PR. This prevents replacing the phi node with its only value if that would lead to invalid IR.
2. Avoid handling unreachable blocks. This was the second approach I pushed. However, this would require some analysis (especially since DT should not be used during execution of the pass, as @nikic mentioned [here](https://github.com/llvm/llvm-project/issues/79175#issuecomment-1906813122). Since @arsenm felt that would be too much. I tried to figure out another way without trying to avoid working on blocks that became unreachable during the execution of the pass.
3. Avoid creating invalid IR at an early stage by avoiding deletion of loop pre-headers whenever they define instructions used by phi nodes in the loop header if removing these values from the phi nodes could lead to the phi node being removed and thus creating self-references (see the code for the details of the checks).

ro-i committed:
fix formatting

ro-i committed:
more precise naming

ro-i commented:
Edit: My terminology wasn't precise. I actually talked about a loop predecessor block, which in my test case is also a pre-header, but in general, we cannot determine whether it's only a loop predecessor or pre-header without further analysis (which I assume shouldn't be done here).

Meinersbur commented:
I am assuming this is what cases the failure: https://github.com/llvm/llvm-project/blob/eef5ea0c42fc07ef2c948be59b57d0df8ec801ca/llvm/lib/IR/Verifier.cpp#L5150-L5151

Having looked closer into it, I think the fault is in the Verifier. Since dominance is not defined in unreachable code, checking dominance is explictly neutred: https://github.com/llvm/llvm-project/blob/b00ad366323cee324b0294f6395c33ae4b047e2c/llvm/include/llvm/Support/GenericDomTree.h#L474-L475
An instruction referencing itself is a special case of violating dominance. Hence, it should be skipped in unreachable code.

Alternatively, InstSimplify needs to be changed in that before replacing an PHIInst (which is allowed to reference itself **IF** the incoming edge is dominated by the PHI -- the check that is skipped in unreachable code) with its only value does not cause any other instruction to reference itself (which is not allowed without exception for unreachable code).

In any case, this is not a problem with JumpThreading and there might also be cases in SimplifyCFG that trigger this. For instance, JumpThreading could just replace the instruction in the BB with an UnreachableInst instead of deleting it. That would just toss the same problem to SimplifyCFG.

ro-i commented:
Hm, I see that this issue might be even larger :D But I'm not sure whether I completely get what you're saying, @Meinersbur .

- In the initial state of the IR function, no instruction is referencing itself. (And there are only statically unreachable blocks because the test case has already been simplified by llvm-reduce.) The IR should be valid, as far as I can tell.
- The issue I see is that the JumpThreading pass does not properly recognize blocks that became unreachable *by actions of the pass itself*, so during the execution of the pass. The pass only checks for empty predecessors, which is not enough. Of course, it cannot use the DominatorTree, since that would be invalid during execution of the pass. **But**, as you pointed out in our other conversation, I could make use of `EliminateUnreachableBlocks` in `BasicBlockUtils`. This would mitigate [what Matt said above](https://github.com/llvm/llvm-project/pull/129501#discussion_r1981633595) because I would not reimplement the logic myself anymore :)
- Another problem is that there are quite some tests for the JumpThreading pass that rely on the fact that the pass used to work on unreachable code instead of ignoring it. See, for example, https://github.com/llvm/llvm-project/blob/eef5ea0c42fc07ef2c948be59b57d0df8ec801ca/llvm/test/Transforms/JumpThreading/ne-undef.ll. The pass should determine that this boils down to a direct jump from bb to bb13, so bb1 would become unreachable directly after handling bb. And there are many such cases in the tests for this pass. I don't understand why, tbh.

Meinersbur commented:
I didn't see that its throug the actions of JumpThreading. Where exactly does crash occur. Within DeleteDeadBlock? Why does it try to fix dead BBs? At the verifier afterwards? After InstCombine on it?

Simplest fix would be to not to try to cleanup unreachable code in JumpThreading. Just let ADCE/SimplifyCFG handle it. Or call EliminateUnreachableBlocks at the end. That would even be cheaper than `DeleteDeadBlocks` since each invocation calls `detachDeadBlocks` which also iterates over all BBs.

There is comment saying: 
> We must remove BB to prevent invalid IR.
Which is AFAIK wrong. Unreachable code in itself is not invalid IR.

ro-i committed:
Revert "more precise naming"

This reverts commit 24a1fca2636d647a697148ba8ad296dc2acb38c1.

ro-i committed:
Revert "fix formatting"

This reverts commit 573c2a98badf712b6157e7cffb6d8f5a2455d2e7.

ro-i committed:
Revert "implement another approach; reset existing tests"

This reverts commit 35aa07dbad7731d0ce52d3b6045db9f29d106053.

ro-i committed:
Revert "replace undef"

This reverts commit e0ccb61acbb90ca486be6d0b89aa18602febcc82.

ro-i committed:
Revert "implement another approach"

This reverts commit d9e764ffdf4c18dcae3e7589d0cc56a8350428b7.

ro-i committed:
Revert "implement feedback"

This reverts commit c97a6c65c3935653d8431daf31976c31266ab38f.

ro-i committed:
Revert "[IR] Avoid self-referencing values caused by PHI node removal"

This reverts commit b1b2f0acbcee984e8e085df77f6b117b05b6406b.

ro-i committed:
[IR][JumpThreading] Fix infinite recursion on compare self-reference

In unreachable code, constant PHI nodes may appear and be replaced by
their single value. As a result, instructions may become
self-referencing. This commit adds checks to avoid going into infinite
recursion when handling self-referencing compare instructions in
`evaluateOnPredecessorEdge()`.

ro-i committed:
fix formatting

nikic commented on llvm/lib/Transforms/Scalar/JumpThreading.cpp:
```diff
@@ -1519,12 +1519,20 @@ Constant *JumpThreadingPass::evaluateOnPredecessorEdge(BasicBlock *BB,
-          evaluateOnPredecessorEdge(BB, PredPredBB, CondCmp->getOperand(0), DL);
-      Constant *Op1 =
-          evaluateOnPredecessorEdge(BB, PredPredBB, CondCmp->getOperand(1), DL);
+      Constant *Op0 = CondCmp->getOperand(0) == CondCmp
```
This handles one specific kind of recursion, but there may be recursion across multiple instructions. It would be better to add a Visited set instead.

ro-i replied:
You mean directly in `evaluateOnPredecessorEdge`?

ro-i replied:
We could do something like this (see the commit I just pushed), I guess

ro-i commented:
After talking about this again with @Meinersbur , I think I got a better view on this issue.
I initially didn't really understand your comment [here](https://github.com/llvm/llvm-project/pull/129501#pullrequestreview-2655373280), @nikic . Sorry about that. I was a bit confused because the pass itself has this comment about requiring the removal of unreachable blocks to prevent invalid IR. So I thought that we must either prevent invalid IR or remove unreachable blocks properly. But I now see that it's ok to just handle invalid IR and pass it along (it will then fail at a later stage).

ro-i committed:
implement feedback

ro-i committed:
public -> private

Meinersbur commented:
When using a `Visited` list to stop infinite recursion, the list must be reverted to the previous state when leaving each level. Think of two `icmp` instructions referencing the same PHI (in reachable code!). The `evaluateOnPredecessorEdge` on the first icmp will mark the PHI as visited, the second would think the PHI is recursive since it was already visited.

Could it be sufficient to check `CondComp->getOperand(0) == CondComp` /  `CondComp->getOperand(1) == CondComp` at some point? If iteratively replacing values with identical values, it should evantually replace everything in the chain with the same value. Also reminds me of [finding a loop lin a singly-linked list](https://stackoverflow.com/questions/10275587/finding-loop-in-a-singly-linked-list#).

ro-i commented:
Hm, I'm not 100% sure about the issue. I'm currently using a set to track all visited values when visiting an instruction. As soon as another instruction is visited, there will be a new set. Am I missing something?

ro-i commented:
```
%x = icmp %a, %p
%y = icmp %x, %p
```
When handling %y, `evaluateOnPredecessorEdge` will descend into %x. There, it will already handle %p. Consequently, %p will already be in the Visited set when it should be handled as the second argument to %y.
That's what you meant, @Meinersbur , right?

arsenm reviewed (commented):
This version seems plausible to me but I haven't looked in detail

ro-i committed:
implemented feedback

ro-i commented:
Added a fix and a testcase for the issue @Meinersbur mentioned.

(@arsenm yeah, ig this approach does make sense. From what I've read in past discussions, others think that the pass needs a rewrite, but I'm glad it does not seem necessary at the moment and that I can fix the current issue in a more straightforward way :) )

ro-i commented:
But `evaluateOnPredecessorEdge` recurses only into `CmpInst`s if I'm not mistaken?

When I run your example:
```
define void @foo() {
Preheader:
        br label %LoopHeader

LoopHeader:
        %common_phi = phi i32 [0, %Preheader], [%phi1, %Latch]
        br label %Latch

Latch:
        %phi1 = phi i32 [%common_phi, %LoopHeader]
        %phi2 = phi i32 [%common_phi, %LoopHeader]
        %cmp = icmp eq i32 %phi1, %phi2
        br i1 %cmp, label %LoopHeader, label %Exit

Exit:
        ret void
}
```
I get:
```
define void @foo() {
Preheader:
  br label %Latch

Latch:                                            ; preds = %Preheader, %Latch
  %common_phi = phi i32 [ 0, %Preheader ], [ %common_phi, %Latch ]
  %cmp = icmp eq i32 %common_phi, %common_phi
  br i1 %cmp, label %Latch, label %Exit

Exit:                                             ; preds = %Latch
  ret void
}
```
which should be what you intended?

Meinersbur reviewed (commented):
I overlooked the responses after https://github.com/llvm/llvm-project/pull/129501#issuecomment-2748329091. So yes, de6ae293645a90bf76d7a0c74ffff6b76b2d9729 addresses my concern. The PHI was an example, I did not look into what `evaluateOnPredecessorEdge` recurses over.

I would prefer if the `insert` and `erase` would be done in the same call. A final review by @nikic would be good.

ro-i committed:
implement feedback

ro-i commented:
> I would prefer if the insert and erase would be done in the same call.

implemented

arsenm commented on llvm/lib/Transforms/Scalar/JumpThreading.cpp:
```diff
@@ -1537,10 +1535,12 @@ Constant *JumpThreadingPass::evaluateOnPredecessorEdge(
         Visited.erase(V0);
       }
       if (Value *V1 = CondCmp->getOperand(1); !Visited.contains(V1)) {
+        Visited.insert(V1);
```
can you use the iterator here to erase after

nikic commented on llvm/lib/Transforms/Scalar/JumpThreading.cpp:
```diff
@@ -1519,12 +1526,24 @@ Constant *JumpThreadingPass::evaluateOnPredecessorEdge(BasicBlock *BB,
+      Constant *Op0 = nullptr;
+      Constant *Op1 = nullptr;
+      if (Value *V0 = CondCmp->getOperand(0); !Visited.contains(V0)) {
+        Visited.insert(V0);
```
I think it would be cleaner to do these checks in evaluateOnPredecessorEdge(), not at each call to evaluateOnPredecessorEdge. Doesn't really matter right now, but will make sure this keeps getting handled if this code handles more than icmp in the future.

Please also combine the contains and insert calls by using the return value of insert.

ro-i replied:
> I think it would be cleaner to do these checks in evaluateOnPredecessorEdge(), not at each call to evaluateOnPredecessorEdge. Doesn't really matter right now, but will make sure this keeps getting handled if this code handles more than icmp in the future.

I adapted the code to insert and double check at the start of evaluateOnPredecessorEdge, see my latest commit. However, there still needs to be a `contains` before each call because, otherwise, the caller doesn't know whether it's allowed to erase the Value after the call.

nikic replied:
You can also handle the erase in the same place, e.g. by adding a `auto _ = make_scope_exit([V]() { Visited.erase(V); });`

ro-i replied:
I didn't know `make_scope_exit`, thanks!

nikic reviewed (commented):
Approach looks fine now

ro-i committed:
insert and double check at function start

ro-i committed:
use make_scope_exit

nikic reviewed (approved):
LGTM
